src/gordon_gcp/clients/http.py

```diff
@@ -95,15 +92,96 @@ def __init__(self, auth_client=None, session=None):
+                retry_type is "custom".  Must return True for "continue retrying
+                (subject to tries and timeout)" or False for "stop retrying".
+                Last argument must be the response from _request().  If the
+                function takes any other arguments, their values must be
+                supplied in predicate_args and/or predicate_kwargs.
+            predicate_args (list): (optional) See retry_predicate.
+            predicate_kwargs (dict): (optional) See retry_predicate.
+            extra_401_retry (bool): (optional) If True, run one more request
+                attempt on a 401 (unauthorized), regardless of the value of
+                retry_type.  This makes sure we've tried generating a new token.
         Returns:
             (str) HTTP response body.
         Raises:
-            :exc:`.GCPHTTPError`: if any exception occurred.
+            :exc:`.GCPHTTPError`: If a response code >=400 was received and
+                there are no more tries (or no retrying is specified).
         """
-        refresh_attempt = kwargs.pop('cred_refresh_attempt', 0)
+        retry_predicates = {
+            "none": lambda resp: False,  # never retry
+            # retry until success, subject to tries/timeout
+            "simple": lambda resp: resp.status >= 400,
+            # don't retry on 4xx
+            "no4xx": lambda resp: resp.status >= 500,
+            "custom": retry_predicate,
+        }
+        actual_predicate = retry_predicates[retry_type]
+        if predicate_args is None:
+            predicate_args = []
+        if predicate_kwargs is None:
+            predicate_kwargs = {}
+        predicate_partial = partial(
+            actual_predicate, *predicate_args, **predicate_kwargs
+        )
 
+        def backoff_logger(details):
+            """See https://github.com/litl/backoff#event-handlers."""
+            logging.info(_utils.REQ_LOG_FMT.format(method, url) +
+                         " -> Failed predicate, trying again "
+                         f"({details.tries} tries so far).")
+
+        def giveup_logger(details):
+            """See https://github.com/litl/backoff#event-handlers."""
+            logging.info(_utils.REQ_LOG_FMT.format(method, url) +
+                         " -> Failed predicate, but no more retries "
+                         f"({details.tries} tries so far).")
+
+        resp = await backoff.on_predicate(
+            backoff.expo, max_tries=tries, max_time=timeout,
+            predicate=predicate_partial, on_backoff=backoff_logger,
+            on_giveup=giveup_logger,
+        )(self._request)(method, url, params, body, headers)
+        # one extra try on a 401 to make sure we've refreshed the token
+        if resp.status == HTTPStatus.UNAUTHORIZED and extra_401_retry:
+            logging.info(_utils.REQ_LOG_FMT.format(method, url) +
+                         " -> Unauthorized.  Trying one more time with new "
+                         "token.")
```

Comment:
What's the reasoning for changing the log message?

Reply:
Also, the style guide uses single quotes for strings.

Reply:
Just making it more consistent with the others.  I can change it back if you hate it.

Am I missing something in the style guide?  What I saw says 'use either one as long as you're consistent' - which this isn't, so I'll change it, but I want to make sure I understand.